keps/sig-network/4963-kube-proxy-flowtables-fastpath/README.md

```diff
@@ -0,0 +1,463 @@
+  |_____|   |                     flowtable                             |
+     |      |                                                           |
+    / \     |                                                           |
+   /hit\_no_|                                                           |
+   \ ? /                                                                |
+    \ /                                                                 |
+     |__yes_________________fastpath bypass ____________________________|
+
+             Fig.1 Netfilter hooks and flowtable interactions
+```
+
+Enabling the flowtable fastpath requires to use nftables and only needs to create a flowtable on the corresponding network interface and add one rule in the forward chain.
+
+Example configuration:
+
+```
+table inet x {
+        flowtable f {
+                hook ingress priority 0; devices = { eth0, eth1 };
+        }
+        chain y {
+                type filter hook forward priority 0; policy accept;
+                ip protocol tcp flow add @f
+                counter packets 0 bytes 0
+        }
+}
+```
+
+Kube-proxy will allow users to opt-in for Service acceleration, since it is not possible to guarantee that a feature that completely bypass the kernel has unintended consequences in some environments or incompatibility with other networking components.
+
+The opt-in behavior will be based on network interface names and attributes and will use CEL expressions, similar to how we use `--pod-interface-name-prefix` in [KEP-2450](https://github.com/kubernetes/enhancements/tree/master/keps/sig-network/2450-Remove-knowledge-of-pod-cluster-CIDR-from-iptables-rules). The powerful matching expression allow users to deal with more complex environments: nodes with multiple interfaces and only some of them with hardware offload, matches on alias to select only some specific kind of pods, ...
+
+Some example expressions are:
+
+- `interface.name.startsWith(\"eth\")` selects all interfaces with prefix `eth`
+- `interface.type == \"veth\"` selects all interfaces with type `veth`
+- `interface.alias == \"offload\"` selects all interfaces with alias set to `offload`
+
+This expressions can also be concatenated, see https://github.com/google/cel-go for more information.
+
+Kernel bypassing will only be applied once the connection is ESTABLISHED, the rationale behind this is that, once the connection is accelerated, it bypass the network stack, by waiting for the connection to be established we ensure the integration with other network applications (network policies, service meshes, ... ) in the node and we avoid possible performance problems or possible DOS attacks by trying to accelerate every single connection on a node.
+
+### User Stories (Optional)
+
+#### Story 1
+
+I want to use Service traffic acceleration with kube-proxy to improve the performance of my applications.
+
+### Risks and Mitigations
```

Comment:
> I think we should really consider having this be opt-in by Service, probably by adding some way to indicate the expected characteristics of the Service traffic. That also conveniently answers the question of "why are you only doing this for Service traffic?"

This is an interesting approach, if we make it opt-in by service we can remove the opt-in by interface, right?
Any idea or thoughts you have in mind?

Reply:
> Any idea or thoughts you have in mind?

some sort of hint in ServiceSpec about what sort of traffic the Service expects (eg, long-lived vs short-lived, high-bandwidth vs low-bandwidth, quick-response vs slow-response, etc). Or maybe just an enum containing specific connection types we know about specific optimizations for... https://github.com/kubernetes/kubernetes/issues/127259#issuecomment-2436201258

> if we make it opt-in by service we can remove the opt-in by interface, right?

Probably? We really need to understand what the implications are of adding every interface on the system to the flowtable...

Reply:
Update the KEP with clarifications, Service is not the right interface because the Service does not provide hints about how clients will consume it, i.e. I can have a web server behind a service with different URLs and some of them may be short lived and other large lived.

Same as Service is not the right interface,  network interfaces are not, since are going to be confusing, and complex.

I was talking with Pablo (netfilter maintainer) about the UX and the number of interfaces:
- UX, he will add some patches to improve the problems with short lived flows, and he acks having an option to offload based on the type of connection is a good idea
- interfaces. he mentioned PhilSutter is working in a wildcard option to choose interfaces, so adding all interfaces to the flowtable does not look problematic.

Reply:
expanded to clarify this